ZEPPELIN-189: Add Apache Geode Interpreter for Zeppelin

tzolov opened the pull request:
Syntax look like this:
%geode.oql
SELECT * FROM /regionOne o WHERE o.field1 > 146 AND o.field2 LIKE '%some%';

bzz commented on geode/pom.xml:
```diff
@@ -0,0 +1,167 @@
+      </plugin>
+    </plugins>
+  </build>
+	<!--   
```
Is this repository still required?

bzz commented on geode/pom.xml:
```diff
@@ -0,0 +1,167 @@
+	<!--   
+	<repositories>
+		<repository>
+			<id>spring-gemstone-releases</id>
```
Is this repository still required?

bzz commented on geode/pom.xml:
```diff
@@ -0,0 +1,167 @@
+	<!-- 	
+	<dependency>
+		<groupId>com.gemstone.gemfire</groupId>
+		<artifactId>gemfire</artifactId>
```
Is this dependency still needed?

tzolov replied:
removed as well

bzz commented:
Can you double check formatting of pom?
Except for that it looks good to me, the only thing is that discussion on the new interpreters contributions is still ongoing on the list dev@

tzolov commented:
@bzz, thanks for reviewing this! With the last commits, I've removed the obsoleted dependencies and cleaned the pom format. 
@buzz, @Leemoonsoo, could you please point me to the @dev discussion concerning the interpreter contribution?

Leemoonsoo commented:
@tzolov 
The subject is 'interpreter with no test'. But the discussion converged to managing interpreters.
http://apache-zeppelin-incubating-dev-mailing-list.75694.x6.nabble.com/Interpreter-with-no-test-td430.html

bzz commented:
Great! 
One more thing @tzolov just to double-check, are you sure 96ebdef belongs to this PR?

tzolov commented:
Ah, thanks for spotting this @bzz ! 
Not sure how i've managed to do it but 96ebdef  is definitely not one of my commits and it is not related to ZEPPELIN-189. 
Would it be safe if i remove the commit from the history (e.g. git rebase -i HEAD~XX) and force push again?

bzz commented:
Sure! Yes, I think that is safe to do.

martin-g commented on geode/src/main/java/org/apache/zeppelin/geode/GeodeOqlInterpreter.java:
```diff
@@ -0,0 +1,297 @@
+    logger.info("Geode open connection called!");
+    try {
+      clientCache = getClientCache();
+      queryService = clientCache.getQueryService();
```
Is it possible/allowed to call `#open()` more than once?
Because if it is possible then it will leak resources (ClientCache and QueryService).
Maybe it is a good idea to always start with a call to `close()` before creating new resources?

tzolov replied:
@martin-g , this is a reasonable observation and as a safeguard it should hurt to add it (in my next commit perhaps). 
But yet again if the framework doesn't close the interpreters before opening them, then this would be a larger issue affecting all Zeppelin interpreter implementations.

tzolov commented:
@bzz , the parasite commits are cleaned and the code is rebased to the latest zeppelin/master. 
I've also committed a small change to allow control of the number of results displayed on the frontend: 0ff81c9.

bzz commented:
Thank you.
Looks good to me.

Leemoonsoo commented:
Do you mind add geode interpreter to https://github.com/apache/incubator-zeppelin/blob/master/zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java#L384 ? which will be the default value when zeppelin-site.xml is missing.

tzolov commented:
Sure @Leemoonsoo i've just added it. Thanks for the effort of reviewing it!

Leemoonsoo commented:
Thanks. Looks good to me

tzolov commented:
FYI, I made a small commit just to trigger a new Travis CI build. Sometimes TravisCI fails with no obvious reasons and there is no way to restart it but to commit. 
The change itself is addition of a safeguard in the open() method to ensure the previous connections are closed. (e.g. some sort of method idempotency guarantees).

martin-g commented:
:+1:

bzz commented:
Thank you for contributing!
It have just been merged to master. GH mirror is somehow slow on catching up though

Leemoonsoo commented:
To resolve github mirror sync issue, i used to file an issue to INFRA, like

https://issues.apache.org/jira/browse/INFRA-9974?jql=project%20%3D%20INFRA%20and%20reporter%20%3D%20moon%20and%20text%20~%20sync

tzolov commented:
@bzz, @Leemoonsoo thank you for the support and for the passions  ;)  
@martin-g  thanks for the valuable suggestions? 

@bzz, the GH is still not in sync. Shall i submit an INFRA ticket or you've done it already?

tzolov commented:
here is the INFRA ticket: https://issues.apache.org/jira/browse/INFRA-10087
